src/librustdoc/html/highlight.rs

```diff
@@ -264,6 +265,66 @@ fn empty_line_number(out: &mut impl Write, _: u32, extra: &'static str) {
     out.write_str(extra).unwrap();
 }
 
+fn get_expansion<'a, W: Write>(
+    token_handler: &mut TokenHandler<'_, '_, W>,
+    expanded_codes: Option<&'a Vec<ExpandedCode>>,
+    line: u32,
+) -> Option<&'a ExpandedCode> {
+    if let Some(expanded_codes) = expanded_codes
+        && let Some(expanded_code) = expanded_codes.iter().find(|code| code.start_line == line)
+    {
+        let (closing, reopening) = if let Some(current_class) = token_handler.current_class
+            && let class = current_class.as_html()
+            && !class.is_empty()
+        {
+            ("</span>", format!("<span class=\"{class}\">"))
+        } else {
+            ("", String::new())
+        };
+        let id = format!("expand-{line}");
+        token_handler.pending_elems.push((
+            Cow::Owned(format!(
+                "{closing}\
+<span class=expansion>\
+    <input id={id} type=checkbox>\
+    <label for={id}>↕</label>{reopening}",
```

Comment:
Why is this done differently than the normal expand/collapse system, but in code and in presentation?

  ![image](https://github.com/user-attachments/assets/1f41926f-87fd-4033-bef8-5f5fb7217ea3)

Reply:
Because it would force to duplicate all the common code. For example in:

```
func(arg, macro!(blabla), arg3);
```

Only `macro!(blabla)` will be different, so it's not interesting to duplicate all the rest. Instead, we put the original code in a `span`, the expanded one in another and then we can simply switch them based on the status of the `checked` property of the input. I'm also not a big fan of our current `details` approach as it requires more JS to be working fully as expected but it's a debate for another time.